Add UInt/Int to _str_sizehint for printing to strings

Seelengrab opened the pull request:
The fallback value of 8 bytes is not enough for the majority of UInt/Int arguments (x > 2^26.5), leading to a potential extra allocation in the caller (e.g. `print_to_string`). This PR returns constants (similar to Float64/32) for these two types of a maximum size, to prevent that allocation in performance critical code. Another option for preventing that allocation would be calling `ceil(Int, log10(x))`, though that would be slightly slower (~15ns on my machine, effectively eating up the gains saved by the lack of allocation). It should be noted that the benchmark below is somewhat unstable, as it depends on the numbers generated (though I expect it to be favorable to this PR, since larger integers are more likely to be drawn than those less than 2^26.5)

This is tangentially related to #29550 if the given arguments are large enough and not hardcoded, as well as [this discourse thread](https://discourse.julialang.org/t/help-with-binary-trees-benchmark-games-example/37307/) (which gets sped up by a few tens of ms on my machine, since the inputs are large enough and occur often enough).

I would like to run more benchmarks, seeing if this speeds up more code "in the wild", but it's hard to find examples since the circumstances are rather peculiar.

----

Benchmarks:

without PR
```
julia> @benchmark string("x", "[", x, ",", y, "]") setup=(x=rand(Int);y=rand(Int))          
BenchmarkTools.Trial:                                                                       
  memory estimate:  480 bytes                                                               
  allocs estimate:  10                                                                      
  --------------                                                                            
  minimum time:     432.323 ns (0.00% GC)                                                   
  median time:      495.455 ns (0.00% GC)                                                   
  mean time:        585.380 ns (6.93% GC)                                                   
  maximum time:     18.040 μs (94.78% GC)                                                   
  --------------                                                                            
  samples:          10000                                                                   
  evals/sample:     198                                                                     
```

with PR
```
julia> @benchmark string("x", "[", x, ",", y, "]") setup=(x=rand(Int);y=rand(Int))          
BenchmarkTools.Trial:                                                                       
  memory estimate:  432 bytes                                                               
  allocs estimate:  9                                                                       
  --------------                                                                            
  minimum time:     414.000 ns (0.00% GC)                                                   
  median time:      435.500 ns (0.00% GC)                                                   
  mean time:        505.774 ns (4.69% GC)                                                   
  maximum time:     14.614 μs (93.58% GC)                                                   
  --------------                                                                            
  samples:          10000                                                                   
  evals/sample:     200                                                                                                            
```

Seelengrab committed:
Add UInt/Int to _str_sizehint for printing to strings

KristofferC commented:
Would it make sense to use `ndigits(x) + (x < 0)`?

Seelengrab commented:
I suspect that would fall into the same category as `ceil(Int, log10(x))` - you'd use slightly less memory, at the cost of some runtime. Not calculating anything is faster than calculating and adding, after all. It's kinda tricky to measure this, since for some reason the overhead when calling/benchmarking this function varies wildly between julia sessions, ranging from ~200ns-1µs. I can't quite explain that, but accounting for that, the original version is still a tiny bit faster, in the best case (measured with Revise in the same session) still ~80ns. 

This is only a bandaid/consistency/partial fix anyway - a proper fix would probably include making the string intermediate generated during the subsequent `print(io, x)` (which calls `write(io, string(x))`, which allocates a new string) in `print_to_string` non-allocating.

KristofferC commented:
> I suspect that would fall into the same category as ceil(Int, log10(x)) - you'd use slightly less memory, at the cost of some runtime.

But `ndigits` should be much faster than `log`. And it doesn't pessimize the (small) common case of small integers.

Seelengrab committed:
Use ndigits instead of constant in _str_sizehint

This doesn't penalize smaller ints in terms of memory at the cost of ~80ns runtime

Seelengrab commented:
That's a good point - I guess that's a tradeoff worth paying 80ns for.

Seelengrab commented:
Failure is unrelated, string/printing tests pass. 

> nested task error: getaddrinfo() thread failed to start while requesting https://httpbingo.julialang.org/delay/2?id=50

KristofferC commented on base/strings/io.jl:
```diff
@@ -123,6 +123,10 @@ function _str_sizehint(x)
         return sizeof(x)
     elseif x isa Char
         return ncodeunits(x)
+    elseif x isa UInt || x isa UInt32
```
Could this just be `x isa Base.BitInteger` to catch all the cases?

Seelengrab replied:
I don't think that's worth it - `UInt` and friends don't need the `x < 0` check and are faster without it (I haven't checked, but I think it's because the type determines the branch absolutely, so all other branches get compiled away. If I mix `Int` and `UInt` definitions, that doesn't happen an `UInt` becomes slower). `[U]Int16` has a maximum of 6 characters, which seems odd to grow an array by so feels fine to let it hit the `return 8`.

KristofferC replied:
> don't need the x < 0 check and are faster without it 

I am surprised about that, for example

```
julia> f(x) = x < 0;

julia> @code_llvm f(UInt(3))
;  @ REPL[35]:1 within `f'
define i8 @julia_f_1418(i64 zeroext %0) {
top:
  ret i8 0
}
```

Also if this was not the case, a single assembly instruction should be insignificant compared to e.g. the `ndigits` call, or?

Seelengrab replied:
I do agree that it should not matter, all I'm saying is that I observed a difference 🤷‍♂️ Would be useful if you also ran the benchmark with and without this patch:

```patch
diff --git a/base/strings/io.jl b/base/strings/io.jl
index 7a940e32fc..de09e6ea40 100644                 
--- a/base/strings/io.jl                            
+++ b/base/strings/io.jl                            
@@ -123,9 +123,7 @@ function _str_sizehint(x)       
         return sizeof(x)                           
     elseif x isa Char                              
         return ncodeunits(x)                       
-    elseif x isa UInt || x isa UInt32              
-        return ndigits(x)                          
-    elseif x isa Int || x isa Int32                
+    elseif x isa BitInteger                        
         return ndigits(x) + (x < zero(x))          
     else                                           
         return 8                                   
```

It's kinda hard to benchmark this for me, since the overhead keeps changing and I can't seem to get to the fast 350-400ns space I was in earlier... code layout makes a mess of things, I think.

Seelengrab replied:
New day, new try.

With explicit cases for UInt64/Int64:

```
julia> @benchmark string("x", "[", x, ",", y, "]") setup=(x=rand(UInt);y=rand(UInt)) evals=1
BenchmarkTools.Trial:                                                                       
  memory estimate:  432 bytes                                                               
  allocs estimate:  9                                                                       
  --------------                                                                            
  minimum time:     2.400 μs (0.00% GC)                                                     
  median time:      2.500 μs (0.00% GC)                                                     
  mean time:        2.687 μs (0.00% GC)                                                     
  maximum time:     19.500 μs (0.00% GC)                                                    
  --------------                                                                            
  samples:          10000                                                                   
  evals/sample:     1                                                                       
```

With BitInteger:

```
julia> @benchmark string("x", "[", x, ",", y, "]") setup=(x=rand(UInt);y=rand(UInt)) evals=1
BenchmarkTools.Trial:                                                                       
  memory estimate:  432 bytes                                                               
  allocs estimate:  9                                                                       
  --------------                                                                            
  minimum time:     3.900 μs (0.00% GC)                                                     
  median time:      4.100 μs (0.00% GC)                                                     
  mean time:        4.332 μs (0.00% GC)                                                     
  maximum time:     30.900 μs (0.00% GC)                                                    
  --------------                                                                            
  samples:          10000                                                                   
  evals/sample:     1                                                                                                                              
```

This is consistent across runs. 

There's one thing I can't explain though - the number of allocated bytes sometimes jumps to 448. Some alignment thing somewhere?

Seelengrab replied:
I don't know how to investigate the difference further, I can only assume it's because `BitInteger` is a somewhat large `Union` - in which case, should `UInt128/Int128` be special cased as well? They can use up to 39 character to print after all, much more than the fallback 8.

KristofferC replied:
I don't know either why it would happen. I know this code is quite sensitive to "inference cycles" where if this code calls out into other code that ends up using this function (so you get a cycle), bad things can happen. Maybe it is something along those lines..? Just guessing.

Seelengrab replied:
Hm, I haven't changed anything in terms of code that ran 🤔 I made sure to restart julia and only ran `using Revise, BenchmarkTools; Revise.track(Base); Revise.revise()` before each test with each version.

FWIW, if I don't split them and go with `BitInteger`, it's a regression - here's the same benchmark on my machine for the original code:

```
julia> using Revise, BenchmarkTools; Revise.track(Base); Revise.revise()                    
                                                                                            
julia> @benchmark string("x", "[", x, ",", y, "]") setup=(x=rand(UInt);y=rand(UInt)) evals=1
BenchmarkTools.Trial:                                                                       
  memory estimate:  480 bytes                                                               
  allocs estimate:  10                                                                      
  --------------                                                                            
  minimum time:     2.700 μs (0.00% GC)                                                     
  median time:      3.900 μs (0.00% GC)                                                     
  mean time:        4.580 μs (0.00% GC)                                                     
  maximum time:     77.700 μs (0.00% GC)                                                    
  --------------                                                                            
  samples:          10000                                                                   
  evals/sample:     1                                                                       
```

So I guess having the two cases split really is best, though I see no clear reson why - they at least both keep the allocations from growing.

Seelengrab committed:
Explicitly use UInt64/Int64 instead of UInt/Int

Seelengrab commented:
bump?

Seelengrab commented:
Should I change anything else about this PR or could this be merged? Since there hasn't been any movement here for some time, should I close it since it doesn't seem like it'll move anytime soon?

KristofferC committed:
Merge branch 'master' into print_to_string_sizehint
